Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-174.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0
Sha: 22c61e0d971b2390a497966bcedf0e09bf9c0860
Containerized:False
Number of Error(s): 8
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-15545/HPCC-Platform/roxie/ccd/ccdqueue.cpp:2848:84: error: void value not ignored as it ought to be
/home/centos/smoketest/PR-15545/HPCC-Platform/roxie/ccd/ccdqueue.cpp:2852:211: error: cannot convert \xe2\x80\x98const IpAddress\xe2\x80\x99 to \xe2\x80\x98TokenBucket*\xe2\x80\x99 for argument \xe2\x80\x986\xe2\x80\x99 to \xe2\x80\x98ISendManager* createSendManager(int, int, int, int, int, TokenBucket*, bool)\xe2\x80\x99
make[2]: *** [roxie/ccd/CMakeFiles/ccd.dir/build.make:225: roxie/ccd/CMakeFiles/ccd.dir/ccdqueue.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:16396: roxie/ccd/CMakeFiles/ccd.dir/all] Error 2
make: *** [Makefile:182: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2021-11-03_17-36-19
time logs started
Prep time   :    16 sec (00:00:16)
Build time  :   328 sec (00:05:28)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   344 sec (00:05:44)
time logs end

[Build error:](https://gist.github.com/e31a46f9266244aff60aeda26dafe519)
Build log [Build log](https://gist.github.com/HPCCSmoketest/e31a46f9266244aff60aeda26dafe519#file-result-21-11-03-17-30-21-log)

--- a/roxie/ccd/ccdqueue.cpp
+++ b/roxie/ccd/ccdqueue.cpp
@@ -2845,10 +2845,11 @@ public:
         if (udpResendLostPackets && udpMaxSlotsPerClient > TRACKER_BITS)
             udpMaxSlotsPerClient = TRACKER_BITS;
         unsigned serverFlowPort = topology->getPropInt("@serverFlowPort", CCD_SERVER_FLOW_PORT);
+        bool sendFlowOnDataPort = topology->getPropBool("@sendFlowOnDataPort", true);
         unsigned dataPort = topology->getPropInt("@dataPort", CCD_DATA_PORT);
         unsigned clientFlowPort = topology->getPropInt("@clientFlowPort", CCD_CLIENT_FLOW_PORT);
         receiveManager.setown(createReceiveManager(serverFlowPort, dataPort, clientFlowPort, udpQueueSize, udpMaxSlotsPerClient, encryptionInTransit));
-        sendManager.setown(createSendManager(serverFlowPort, dataPort, clientFlowPort, udpSendQueueSize, fastLaneQueue ? 3 : 2, bucket, encryptionInTransit));
+        sendManager.setown(createSendManager(sendFlowOnDataPort ? dataPort : serverFlowPort, dataPort, clientFlowPort, udpSendQueueSize, fastLaneQueue ? 3 : 2, bucket, encryptionInTransit));
     }
 
 };

--- a/roxie/udplib/udptrr.cpp
+++ b/roxie/udplib/udptrr.cpp
@@ -54,6 +54,9 @@ static unsigned lastFlowPermitsSent = 0;
 static unsigned lastFlowRequestsReceived = 0;
 static unsigned lastDataPacketsReceived = 0;
 
+// The code that redirects flow messages from data socket to flow socket relies on the assumption tested here
+static_assert(sizeof(UdpRequestToSendMsg) < sizeof(UdpPacketHeader), "Expected UDP rts size to be less than packet header");
+
 class CReceiveManager : implements IReceiveManager, public CInterface
 {
     /*
@@ -573,7 +576,8 @@ class CReceiveManager : implements IReceiveManager, public CInterface
     class receive_data : public Thread 
     {
         CReceiveManager &parent;
-        ISocket *receive_socket;
+        ISocket *receive_socket = nullptr;
+        ISocket *selfFlowSocket = nullptr;
         std::atomic<bool> running = { false };
         Semaphore started;
         
@@ -585,6 +589,7 @@ class CReceiveManager : implements IReceiveManager, public CInterface
             if (check_max_socket_read_buffer(ip_buffer) < 0) 
                 throw MakeStringException(ROXIE_UDP_ERROR, "System socket max read buffer is less than %u", ip_buffer);
             receive_socket = ISocket::udp_create(parent.data_port);
+            selfFlowSocket = ISocket::udp_connect(SocketEndpoint(parent.receive_flow_port, myNode.getIpAddress()));
             receive_socket->set_receive_buffer_size(ip_buffer);
             size32_t actualSize = receive_socket->get_receive_buffer_size();
             DBGLOG("UdpReceiver: rcv_data_socket created port=%d requested sockbuffsize=%d actual sockbuffsize=%d", parent.data_port, ip_buffer, actualSize);
@@ -603,8 +608,11 @@ class CReceiveManager : implements IReceiveManager, public CInterface
             running = false;
             if (receive_socket)
                 receive_socket->close();
+            if (selfFlowSocket)
+                selfFlowSocket->close();
             join();
             ::Release(receive_socket);
+            ::Release(selfFlowSocket);
         }
 
         virtual int run() 
@@ -625,7 +633,13 @@ class CReceiveManager : implements IReceiveManager, public CInterface
                 {
                     unsigned int res;
                     b = bufferManager->allocate();
-                    receive_socket->read(b->data, 1, DATA_PAYLOAD, res, 5);
+                    while (true)
+                    {
+                        receive_socket->read(b->data, 1, DATA_PAYLOAD, res, 5);
+                        if (res!=sizeof(UdpRequestToSendMsg))
+                            break;
+                        selfFlowSocket->write(b->data, res);
+                    }
                     dataPacketsReceived++;
                     UdpPacketHeader &hdr = *(UdpPacketHeader *) b->data;
                     assert(hdr.length == res && hdr.length > sizeof(hdr));

